src/pool/cli.c

```diff
@@ -1454,6 +1422,452 @@ dc_pool_query(tse_task_t *task)
+	} else if (rc != 0) {
+		D_ERROR(DF_UUID": failed to fetch pool map: "DF_RC"\n",
+			DP_UUID(pool->dp_pool), DP_RC(rc));
+		goto out;
+	}
+
+	if (out->tmo_op.po_map_version < pool->dp_map_version_known) {
+		/* We are no longer interested in what the server has. */
+		D_DEBUG(DB_MD, DF_UUID": %p: got stale %u < known %u\n",
+			DP_UUID(pool->dp_pool), task,
+			out->tmo_op.po_map_version, pool->dp_map_version_known);
+		reinit = true;
+		goto out;
+	}
+
+	rc = pool_map_create(cb_arg->mrc_map_buf, out->tmo_op.po_map_version,
+			     &map);
+	if (rc != 0) {
+		D_ERROR(DF_UUID": failed to create pool map: "DF_RC"\n",
+			DP_UUID(pool->dp_pool), DP_RC(rc));
+		goto out;
+	}
+
+	rc = dc_pool_map_update(pool, map, out->tmo_op.po_map_version,
+				false /* connect */);
+
+out:
+	destroy_map_refresh_rpc(cb_arg->mrc_rpc, cb_arg->mrc_map_buf);
+
+	if (reinit) {
+		uint32_t	backoff;
+		int		rc_tmp;
+
+		backoff = d_backoff_seq_next(&arg->mra_backoff_seq);
+		rc_tmp = tse_task_reinit_with_delay(task, backoff);
+		if (rc_tmp == 0) {
+			D_DEBUG(DB_MD,
+				DF_UUID": %p: reinitialized due to "DF_RC
+				" with backoff %u\n", DP_UUID(pool->dp_pool),
+				task, DP_RC(rc), backoff);
+			rc = 0;
+		} else {
+			D_ERROR(DF_UUID": failed to reinitialize pool map "
+				"refresh task: "DF_RC"\n",
+				DP_UUID(pool->dp_pool), DP_RC(rc));
+			if (rc == 0)
+				rc = rc_tmp;
+			reinit = false;
+		}
+	}
+
+	if (!reinit) {
+		D_ASSERTF(pool->dp_map_task == task, "%p == %p\n",
+			  pool->dp_map_task, task);
+		tse_task_decref(pool->dp_map_task);
+		pool->dp_map_task = NULL;
+	}
+
+	D_RWLOCK_UNLOCK(&pool->dp_map_lock);
+
+	if (!reinit) {
+		d_backoff_seq_fini(&arg->mra_backoff_seq);
+		dc_pool_put(arg->mra_pool);
+	}
+
+	return rc;
+}
+
+static int
+map_refresh(tse_task_t *task)
+{
+	struct map_refresh_arg	       *arg;
+	struct dc_pool		       *pool;
+	d_rank_t			rank;
+	unsigned int			version;
+	crt_rpc_t		       *rpc;
+	struct map_refresh_cb_arg	cb_arg;
+	int				rc;
+
+	arg = tse_task_buf_embedded(task, sizeof(*arg));
+	pool = arg->mra_pool;
+
+	if (arg->mra_passive) {
+		/*
+		 * Passive pool map refresh tasks do nothing besides waiting
+		 * for the active one to complete. They avoid complexities like
+		 * whether a dc_pool_create_map_refresh_task caller should
+		 * schedule the resulting task or not and how the caller would
+		 * register its completion callback to the bottom of the
+		 * resulting task's callback stack.
+		 */
+		D_DEBUG(DB_MD, DF_UUID": %p: passive done\n",
+			DP_UUID(pool->dp_pool), task);
+		rc = 0;
+		goto out_task;
+	}
+
+	D_RWLOCK_WRLOCK(&pool->dp_map_lock);
+
+	/* Update the highest known pool map version in all cases. */
+	if (pool->dp_map_version_known < arg->mra_map_version)
+		pool->dp_map_version_known = arg->mra_map_version;
+
+	if (arg->mra_map_version != 0 && !map_known_stale(pool)) {
+		D_RWLOCK_UNLOCK(&pool->dp_map_lock);
+		rc = 0;
+		goto out_task;
+	}
+
+	if (pool->dp_map_task != NULL && pool->dp_map_task != task) {
```

Comment:
So this only happened if there multiple threads within one daos client process?

Reply:
Correct. If an application does concurrent I/Os, it may hit this. One of daos_test cases actually hit this condition in my testing.